src/controllers/plugins/LuaAPI.cpp

```diff
@@ -63,195 +74,90 @@ QDebug qdebugStreamForLogLevel(lua::api::LogLevel lvl)
-
-        lua_pushnil(coro);
-        lua_setfield(coro, LUA_REGISTRYINDEX, name.toStdString().c_str());
-        if (lua_gettop(coro) != 0)
-        {
-            stackDump(coro,
-                      pl->id +
-                          ": timer returned a value, this shouldn't happen "
-                          "and is probably a plugin bug");
-        }
-    });
-    stackDump(L, "before setfield");
-    lua_setfield(L, LUA_REGISTRYINDEX, name.toStdString().c_str());
-    lua_xmove(L, coro, 1);  // move function to thread
-    timer->start();
 
-    return 0;
+    sol::state_view main = sol::main_thread(L);
+
+    sol::thread thread = sol::thread::create(main);
+    sol::protected_function cb(thread.state(), callback);
+    main.registry()[name.toStdString()] = thread;
+
+    QObject::connect(
+        timer, &QTimer::timeout,
+        [pl = L.plugin(), name, timer, cb, thread, main]() {
+            timer->deleteLater();
+            pl->removeTimeout(timer);
+            sol::protected_function_result res = cb();
+
+            if (res.return_count() != 0)
+            {
+                stackDump(thread.lua_state(),
+                          pl->id +
+                              ": timer returned a value, this shouldn't happen "
+                              "and is probably a plugin bug");
+            }
+            main.registry()[name.toStdString()] = sol::nil;
+        });
+    timer->start();
 }
 
-int g_load(lua_State *L)
+sol::variadic_results g_load(ThisPluginState s, sol::object data)
 {
 #    ifdef NDEBUG
-    luaL_error(L, "load() is only usable in debug mode");
-    return 0;
+    (void)data;
+    (void)s;
+    throw std::runtime_error("load() is only usable in debug mode");
 #    else
-    auto countArgs = lua_gettop(L);
-    QByteArray data;
-    if (lua::peek(L, &data, 1))
-    {
-        auto *utf8 = QTextCodec::codecForName("UTF-8");
-        QTextCodec::ConverterState state;
-        utf8->toUnicode(data.constData(), data.size(), &state);
-        if (state.invalidChars != 0)
-        {
-            luaL_error(L, "invalid utf-8 in load() is not allowed");
-            return 0;
-        }
-    }
-    else
-    {
-        luaL_error(L, "using reader function in load() is not allowed");
-        return 0;
-    }
-
-    for (int i = 0; i < countArgs; i++)
-    {
-        lua_seti(L, LUA_REGISTRYINDEX, i);
-    }
-
-    // fetch load and call it
-    lua_getfield(L, LUA_REGISTRYINDEX, "real_load");
-
-    for (int i = 0; i < countArgs; i++)
-    {
-        lua_geti(L, LUA_REGISTRYINDEX, i);
-        lua_pushnil(L);
-        lua_seti(L, LUA_REGISTRYINDEX, i);
-    }
 
-    lua_call(L, countArgs, LUA_MULTRET);
-
-    return lua_gettop(L);
+    // If you're modifying this PLEASE verify it works, Sol is very annoying about serialization
+    // - Mm2PL
+    sol::state_view lua(s);
+    auto load = lua.registry()["real_load"];
+    sol::protected_function_result ret = load(data, "=(load)", "t");
+    return ret;
```

Comment:
If we have this comment, we should test it. Maybe a test that's only in debug mode?

Reply:
Added a comment on the function for this